### PR TITLE
java-spring-boot2: publish application in layers

### LINKED ISSUE #527: Spring Boot 2 stack should switch from fat jar to extracted layers
**Is your feature request related to a problem? Please describe.**

The `Dockerfile` generated, as attached, sounds like `appsody` doesn't follow the best practices while building the Docker image for, as least Spring Boot applications, as `appsody` simply builds a fat jar and load it in.

```
FROM kabanero/java-spring-boot2:0.3 as compile

#setup project folder for java build step
COPY . /project

WORKDIR /project/user-app

USER root

RUN /project/util/check_version build \
 && /project/java-spring-boot2-build.sh package

USER java_user

####
FROM kabanero/ubi8-openjdk:0.3.0

ARG artifactId=appsody-spring
ARG version=1.0-SNAPSHOT
ENV JVM_ARGS=""

LABEL maintainer="IBM Java Engineering at IBM Cloud"
LABEL org.opencontainers.image.version=${version}
LABEL org.opencontainers.image.title=${artifactId}
LABEL appsody.stack="kabanero/java-spring-boot2:0.3.13"

COPY --from=compile /project/user-app/target/app.jar /app.jar

EXPOSE 8080
EXPOSE 8443

ENTRYPOINT [ "sh", "-c", "exec java $JVM_ARGS -Djava.security.egd=file:/dev/./urandom -jar /app.jar" ]
```

**Describe the solution you'd like**

As recommended by Spring community, we should take advantage of the clean separation between dependencies and application resources in a Spring Boot fat jar file.
So we should consider something like this while compiling `Dockerfile`:

```
FROM kabanero/ubi8-openjdk:0.3.0
VOLUME /tmp
ARG DEPENDENCY=target/dependency
COPY ${DEPENDENCY}/BOOT-INF/lib /app/lib
COPY ${DEPENDENCY}/META-INF /app/META-INF
COPY ${DEPENDENCY}/BOOT-INF/classes /app
ENTRYPOINT ["java","-cp","app:app/lib/*","-jar","/app.jar"]
```

> Ref: https://spring.io/guides/gs/spring-boot-docker/

What do you think?

### PR DESCRIPTION
**Pull request by ebullient**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

The stack's build instructions were modified to extract dependencies at build time. The final docker image is then constructed using those extracted dependencies.

The root vs. java user has been cleaned up in both the stack and final docker images.

### Related Issues:

Fixes #527

### CODE REVIEW

File: incubator/java-spring-boot2/image/Dockerfile-stack [hunk 1]
```diff
@@ -7,17 +7,25 @@ RUN  apt-get -qq update \
...
+COPY ./LICENSE /licenses/
 COPY ./project /project
 COPY ./config /config
 COPY ./mvn-stack-settings.xml /usr/share/maven/conf/settings.xml
 COPY ./mvn-stack-settings.xml /project/mvn-stack-settings.xml
 
+RUN groupadd --gid 1000 java_group \
+ && useradd --uid 1000 --gid java_group --shell /bin/bash --create-home java_user \
+ && mkdir -p /mvn/repository \
+ && chown -R java_user:java_group /project \
```
Comment (Kamran64) [inline incubator/java-spring-boot2/image/Dockerfile-stack]: If we're using the `chown` flag, please could you add a `docker-version` constraint on the `requirements` section in the `stack.yaml`. Should be a value of ">= 17.09.0"

Comment (ebullient) [reply incubator/java-spring-boot2/image/Dockerfile-stack]: I'd avoided the chown flag in my first PR. And yea.. I read him the riot act about the same thing in Slack. :-P

Comment (ebullient) [reply incubator/java-spring-boot2/image/Dockerfile-stack]: In the snip above, that is the chown command, not the chown flag on copy...

**Comment by Kamran64** [comment]
Just a question: Do we need both this PR and https://github.com/appsody/stacks/pull/543? I see #543 is extending parts of this PR so I wasn't sure if the intention was that this PR gets merged followed by #543 or we would keep one or the other but just wanted to clarify.

**Comment by ebullient** [comment]
> Just a question: Do we need both this PR and #543? I see #543 is extending parts of this PR so I wasn't sure if the intention was that this PR gets merged followed by #543 or we would keep one or the other but just wanted to clarify.

This PR _was_ ready to go, but then Ozzy started building on it. I can't go back and make it a draft. ;) 
I will close this one, as we'll probably merge Ozzy's once we're happy with it.

**Pull request closed without merging**